### PR TITLE
Restructure the way we render the feed + add billboard locations on signed in feed

### PR DESCRIPTION
**Pull request by Ridhwana**
The initial PR was https://github.com/forem/forem/pull/19533 which we merged and then closed because it was erroring out on the ad blocks.

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

The current home feed loads async and so in this PR we want the billboards to load up async as well. There should be no "content jump", where the billboard or the feed cards show up before the other and create a jump. The feed should stay in "loading" state until all are rendered. 

Hence, in order to do this we've moved the logic around to load up the feed items and set them in organized way in the data (in `Feed.jsx`) before rendering it in the view. 

We implement the following organization for the feed:
```
1. Place the pinned post first
2. Place the image post next 
3. Place the podcast episodes of a podcast that you follow which has new episode out today
4. Place the rest of the stories for the feed
5. Insert the billboards in that array accordingly
- feed_first: Before all home page posts
- feed_second: Between 2nd and 3rd posts in the feed
- feed_third: Between 7th and 8th posts in the feed
```

I've not yet implemented:
- The billboard tag list should accept the user's followed tags for targeting.
- Let's make this work for in_house and external type billboards, while keeping community type as a future discussion for how it might logically target the followers of the billboard organization — however, let's leave this unimplemented 

### Some context: 
The way that the feed was being rendered made it difficult to insert billboards because of the way that we were organising the feed in the render function (i.e. the view). 

In the render function of `homePage.jsx` we were ordering the items on the feed. We then needed to insert the billboards in the following places: 
- `feed_first` - Before all home page posts
- `feed_second` - Between 2nd and 3rd posts in the feed
- `feed_third` - Between 7th and 8th posts in the feed

It became difficult to do so because the placement of the `feed_second` and `feed_third` depended on whether there was a featured post, a pinned post, or podcast episodes and for each of permutations the billboard would need to be placed accordingly. In my opinion the view was going to become very brittle littered with _many_ really difficult to follow if conditions. 

Hence, I thought that perhaps we could apply a refactored approach where we can organize the feed first through the data thats been passed through and then insert the billboard once all the items are placed in an array. This decoupled the need to know what types of items was in the feed in order to place the billboard, and the billboard placement now simply relies on the position. 

### The future 

1. I took a small step to move the the organization of the feed to its own component and complete the view rendering dynamically. However, I feel like the next step is to move the organization of the feed to the server rather than doing it on the fronted for performance reasons. However, the jump to that architecture was significant especially since there were no proper tests in place + the code involved a big function rather than smaller single responsibility ones. Hence, I felt like it makes more sense to make this change incrementally. This leaves the code in a better place and will make it much easier to move to the backend. I'll add a separate ticket for this.

2. Whilst developing I noticed that the Preact replaceNode implementation that we use [here](https://github.com/forem/forem/blob/a82d94e1dbfb850570579230609671076529f1cf/app/javascript/packs/homePageFeed.jsx#L195-L200) and probably other places is being deprecated. `The signature '(vnode: ComponentChild, parent: ContainerNode, replaceNode?: Element | Text): void' of 'render' is deprecated.ts(6387) index.d.ts(293, 4): The declaration was marked as deprecated. ` Replacement Preact 10+ implementation can be found here:https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c. I'll create an issue to replace it. 

4. I’d like to create a _follow up issue_ to remove the [initializeDisplayAdVisibility initialiser from legacy](https://github.com/forem/forem/blob/main/app/assets/javascripts/initializers/initializeDisplayAdVisibility.js) and  migrate whatever is using it to packs + add some tests for this pack file. However if folks feel strongly that about it being done in this PR, I can try to add it here.  

## Related Tickets & Documents

- Related Issue #https://github.com/forem/forem/issues/19313
- Closes #

## QA Instructions, Screenshots, Recordings

❗️ Its important to do thorough testing on this PR as it changes the implementation of the signed in home feed 

A loom Demo to demo and explain how to test:
https://www.loom.com/share/35c560f9a6ae40038fe0993712fb1bef

Note that you want to make sure that scrolling, changing between timeframes and everything else on the feed works as expected. If you test on Uffizi, you will not be able to probably get podcasts to show up. 

We implement the following organization for the feed:
```
1. Place the pinned post first
2. Place the image post next
5. Place the podcast episodes of a podcast that you follow which has new episode out today
6. Place the rest of the stories for the feed
7. Insert the billboards in that array accordingly
- feed_first: Before all home page posts
- feed_second: Between 2nd and 3rd posts in the feed
- feed_third: Between 7th and 8th posts in the feed
```

In order to test this you want to make sure that you have all the items setup in the list above: 

1. Pin a post 
Find a pinned post:
```
pinned_article = PinnedArticle.get
```
if there isn't one then run: 
```
PinnedArticle.set(Article.first)
```

2. Set a post with an image
```
articles = Articles::Feeds::Basic.new(user: User.find(11), page: @page).more_comments_minimal_weight_randomized.to_a
first_article = articles.first
first_article.update!(main_image: "https://pigment.github.io/fake-logos/logos/medium/color/2.png")
```

3. Setup your podcasts: We want to ensure that we have a followed podcasts that has episodes published today so that they can show up on the feed.

```
podcast = Podcast.first

user = <the user you're logged in with>
user.follow podcast

Podcasts::GetEpisodesWorker.perform_async("podcast_id" => podcast.id, "limit" => 10)
podcast_episode = podcast.podcast_episodes
podcast_episodes.limit(4).update(published_at: DateTime.now)
```

No need to do anything for (4).

5. Add some billboards:

```
DisplayAd.create!(
  name: "#{Faker::Lorem.word} feed_first",
  body_markdown: Faker::Lorem.sentence,
  published: true,
  approved: true,
  placement_area: "feed_first"
)
```

```
DisplayAd.create!(
  name: "#{Faker::Lorem.word} feed_second",
  body_markdown: Faker::Lorem.sentence,
  published: true,
  approved: true,
  placement_area: "feed_second"
)
```

```
DisplayAd.create!(
  name: "#{Faker::Lorem.word} feed_third",
  body_markdown: Faker::Lorem.sentence,
  published: true,
  approved: true,
  placement_area: "feed_third"
)
```

You can update the configurations by removing and adding some items and they should still maintain the order.

### UI accessibility concerns?
None

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [optional] Are there any post deployment tasks we need to perform?
We need to bust the cache on the home feed through admin and monitor the signed in feed once deployed. 

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/javascript/articles/Feed.jsx [hunk 1]
```diff
@@ -3,92 +3,263 @@ import { useEffect, useState } from 'preact/hooks';
...
 
-        setFeedItems(organizedFeedItems);
-      } catch {
-        if (!onError) setOnError(true);
+        if (isHTML(result)) {
+          resolvedValue = await result.value.text();
+        }
+        fullfilledPromises.push(resolvedValue);
+      } else {
+        Honeybadger.notify(result.reason);
```
Comment (Ridhwana) [inline app/javascript/articles/Feed.jsx:239]: @maestromac I wanted to double check with if this is the correct way (i.e. is this enough context) to notify honeybadger with or should I add any extra data?

**Review by jaw6 — APPROVED** [review]
:+1: Error-handling looks right to me, but I'm haven't had much success replicating the original ad-blocking scenario in my local environment

**Review by benhalpern — APPROVED** [review]
Looks great to me!

File: app/javascript/articles/Feed.jsx (hunk 1 above)
Comment (maestromac) [reply app/javascript/articles/Feed.jsx:239]: Hmm what does `result.reason` look like?

It's fine but doesn't hurt if you want to provide a little more context ie
```
    Honeybadger.notify("failed to fetch ...", context: { notification_id: notification.id })
```

**Pull request merged**